packages/toolkit/src/query/react/buildHooks.ts

```diff
@@ -781,35 +782,60 @@ export function buildHooks<Definitions extends EndpointDefinitions>({
   }
 
   function buildMutationHook(name: string): UseMutation<any> {
-    return ({ selectFromResult = defaultMutationStateSelector } = {}) => {
+    return ({
+      selectFromResult = defaultMutationStateSelector,
+      fixedCacheKey,
+    } = {}) => {
       const { select, initiate } = api.endpoints[name] as ApiEndpointMutation<
         MutationDefinition<any, any, any, any, any>,
         Definitions
       >
       const dispatch = useDispatch<ThunkDispatch<any, any, AnyAction>>()
       const [promise, setPromise] = useState<MutationActionCreatorResult<any>>()
 
-      useEffect(() => () => promise?.reset(), [promise])
+      useEffect(
+        () => () => {
+          if (!promise?.arg.fixedCacheKey) {
+            promise?.reset()
+          }
+        },
+        [promise]
+      )
 
       const triggerMutation = useCallback(
         function (arg) {
-          const promise = dispatch(initiate(arg))
+          const promise = dispatch(initiate(arg, { fixedCacheKey }))
           setPromise(promise)
           return promise
         },
-        [dispatch, initiate]
+        [dispatch, initiate, fixedCacheKey]
       )
 
       const { requestId } = promise || {}
       const mutationSelector = useMemo(
         () =>
-          createSelector([select(requestId || skipToken)], selectFromResult),
-        [select, requestId, selectFromResult]
+          createSelector(
+            [select({ fixedCacheKey, requestId: promise?.requestId })],
+            selectFromResult
+          ),
+        [select, promise, selectFromResult, fixedCacheKey]
       )
 
       const currentState = useSelector(mutationSelector, shallowEqual)
-      const originalArgs = promise?.arg.originalArgs
-      const reset = useCallback(() => setPromise(undefined), [])
+      const originalArgs =
+        fixedCacheKey == null ? promise?.arg.originalArgs : undefined
+      const reset = useCallback(() => {
+        if (promise) {
+          setPromise(undefined)
+        } else if (fixedCacheKey) {
+          dispatch(
+            api.internalActions.removeMutationResult({
+              requestId,
+              fixedCacheKey,
+            })
+          )
+        }
+      }, [dispatch, fixedCacheKey, promise, requestId])
```

Comment:
💁

I think this logic makes it so that when using a fixed cache key, the component that calls `reset` needs to call it twice before it actually removes the result for that key

e.g.

- Components A & B with a shared mutation key
- Component A triggers the mutation
- Component A stores the promise in state
- Components A & B both receive the mutation result
- Component A calls `reset`
- The promise state in component A is set to undefined, but the mutation result for the cache key is left alone
- Components A & B both still receive the mutation result
- Component A calls `reset` again
- The mutation result for the cache key is now removed

Reply:
Good catch. My thought was: Setting `promise` to `undefined` should trigger the `useEffect` cleanup further up, which also triggers a `reset`.

But unfortunately, that skips in that case:
```js
      useEffect(
        () => () => {
          if (!promise?.arg.fixedCacheKey) {
            promise?.reset()
          }
        },
        [promise]
      )
```

Could you open a PR against the integration branch and maybe also add a test for this to prevent regressions?

Reply:
yup, no problem

Reply:
Thank you!